### PR TITLE
Fix XPBD elastic multiplier persistence

### LINKED ISSUE #2933: Review XPBD handling of elastic Lagrange multipliers
In the XPBD solver, spring and bending (and, from #2845, tets) kernels not not track Lagrange multipliers over iterations. This means they actually behave like hard constraints in the limit and do not respect the model's material stiffness.

_Originally posted by @gdaviet in https://github.com/newton-physics/newton/pull/2845#discussion_r3228862970_
            

### Confirmation from Codex

In XPBD, the accumulated Lagrange multiplier is what makes compliance represent a finite material stiffness. At convergence of a compliant constraint:

  C(x) + alpha * lambda = 0

  so nonzero constraint error can remain according to the stiffness/compliance. If lambda is reset every iteration, the fixed point instead becomes essentially:

  C(x) = 0

  for equality constraints, so unlimited iterations drive the material toward a hard constraint, not the calibrated finite stiffness.

  In this code:

  - Springs and bending have lambda arrays in the kernels, but they are zeroed inside every solver iteration: newton/_src/solvers/xpbd/solver_xpbd.py:447, newton/
    _src/solvers/xpbd/solver_xpbd.py:468. That defeats the XPBD accumulated-lambda term used in newton/_src/solvers/xpbd/kernels.py:363 and newton/_src/solvers/xpbd/
    kernels.py:460.
  - Tets do not track lambdas in the launched solve_tetrahedra kernel, and the active code comments out tet_materials usage: newton/_src/solvers/xpbd/kernels.py:497.
    It uses soft_body_relaxation as a compliance-like scalar instead: newton/_src/solvers/xpbd/kernels.py:539.

  So your assumption is correct for finite-stiffness elastic behavior: increasing iterations does not converge to the intended material stiffness. It converges
  toward a hard/inextensible projection, while finite iteration counts give an iteration-, timestep-, mass-, and ordering-dependent effective stiffness. Carrying
  lambdas across nonlinear iterations within the timestep is the key missing piece; carrying them across timesteps is optional warm starting, not the core
  correctness issue.

### PR DESCRIPTION
**Pull request by devshahofficial**
## Description

Preserve accumulated XPBD Lagrange multipliers across nonlinear iterations for springs, bending edges, and both tetrahedral material constraints. Tetrahedral relaxation now scales the multiplier and position update together. Differentiable solves write each iteration to fresh multiplier storage so reverse-mode history is preserved.

This keeps finite material compliance stable as iteration counts increase.

Closes #2933.

## Checklist

- [x] New or existing tests cover these changes
- [x] The documentation is up to date with these changes
- [x] `CHANGELOG.md` has been updated (if user-facing change)

## Test plan

```
uv run --extra dev -m newton.tests -k test_solver_xpbd
uvx pre-commit run -a
```

32 focused tests pass on CPU, including spring, bending, and tetrahedron iteration-independence regressions plus multi-iteration reverse-mode coverage.

## Bug fix

**Steps to reproduce:**

1. Build a finite-stiffness spring, bending edge, or tetrahedron.
2. Start it away from its rest configuration.
3. Compare one step with converged low and high nonlinear iteration counts.
4. On main, the higher count drives the elastic constraint toward a hard constraint.

**Minimal reproduction:**

```python
import newton

builder = newton.ModelBuilder()
builder.add_particle((0.0, 0.0, 0.0), mass=0.0)
builder.add_particle((1.0, 0.0, 0.0), mass=1.0)
builder.add_spring(0, 1, ke=10.0, kd=0.0)
model = builder.finalize()
solver = newton.solvers.SolverXPBD(model, iterations=8)
```

## Summary by CodeRabbit

* **Bug Fixes**
  * Fixed XPBD spring, bending, and tetrahedral constraints becoming artificially rigid when nonlinear iteration counts increase.
  * Improved compliance consistency across varying iteration counts for more stable, repeatable simulations.

* **Tests**
  * Added regression tests covering iteration-independent compliance for elastic spring, elastic bending, and elastic tetrahedral constraints.
  * Added a regression test to ensure spring constraint multiplier behavior remains correct under reverse-mode autodiff.

### CODE REVIEW

File: newton/_src/solvers/xpbd/solver_xpbd.py [hunk 1]
```diff
@@ -459,10 +459,15 @@ def step(
...
+                spring_constraint_lambdas = wp.zeros_like(model.spring_rest_length)
             edge_constraint_lambdas = None
             if model.edge_count:
-                edge_constraint_lambdas = wp.empty_like(model.edge_rest_angle)
+                edge_constraint_lambdas = wp.zeros_like(model.edge_rest_angle)
+            tet_constraint_lambdas = None
+            if model.tet_count:
+                tet_constraint_lambdas = wp.zeros(
+                    model.tet_count * 2, dtype=float, device=model.device, requires_grad=requires_grad
+                )
```
[the remark this answers is not in the document]
Comment (devshahofficial) [reply newton/_src/solvers/xpbd/solver_xpbd.py]: @coderabbitai Addressed in a5efabaa. Differentiable solves now allocate a fresh multiplier output per nonlinear iteration; forward-only solves alternate two buffers. I added verify_autograd_array_access coverage and check the spring-stiffness gradient (-1/121), plus explicit non-rigid/non-zero correction assertions for spring, bending, and tet compliance. All 32 focused XPBD tests and pre-commit pass.
